### PR TITLE
Run WASM E2E tests only when relevant changes are made

### PR DESCRIPTION
**Pull request by chatton**
## Description

Marking as R4R to run tests.

closes: #5090

This PR adds an additional workflow that will run only the wasm tests. This workflow runs whenever relevant files are changed.

The WASM tests have been removed from the regular e2e.yml workflow.

This would be possible by building and tagging an image that looks like `ghcr.io/cosmos/ibc-go-wasm-simd:latest`, however we should probably instead also make it so this image does not get hard coded for this test, and is passed correctly. This could be done in a follow up I think.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: .github/workflows/e2e-wasm.yaml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    with:
+      # with each test, we build an image from the current code.
+      build-and-push-docker-image-wasm: true
+      # if the test fails, we upload logs so that we can download them from the UI.
+      upload-logs: true
+      chain-image: ghcr.io/cosmos/ibc-go-wasm-simd
+      # with regular tests, both images are the same.
+      chain-a-tag: '${{ needs.determine-image-tag.outputs.simd-tag }}'
+      chain-b-tag: '${{ needs.determine-image-tag.outputs.simd-tag }}'
+      chain-binary: 'simd'
```
Comment (chatton) [inline .github/workflows/e2e-wasm.yaml:66]: all this is the same as e2e.yml

File: .github/workflows/e2e-wasm.yaml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      # only run the grandpa test suite for wasm tests.
+      test-entry-point: 'TestGrandpaTestSuite'
```
Comment (chatton) [inline .github/workflows/e2e-wasm.yaml:68]: change to run specific test instead of excluding

File: .github/workflows/e2e-wasm.yaml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    uses: ./.github/workflows/e2e-test-workflow-call.yml
+    # unless we explicitly tell the workflow to inherit secrets, required secrets such as GITHUB_TOKEN will not be
+    # provided to the workflow. This would cause privileged operations to fail.
+    secrets: inherit
+    with:
+      # with each test, we build an image from the current code.
+      build-and-push-docker-image-wasm: true
+      # if the test fails, we upload logs so that we can download them from the UI.
+      upload-logs: true
+      chain-image: ghcr.io/cosmos/ibc-go-wasm-simd
```
Comment (chatton) [inline .github/workflows/e2e-wasm.yaml:62]: specify correct image (not actually required yet as the image is hard coded in the test.)

File: .github/workflows/e2e-wasm.yaml [hunk 4]
```diff
@@ -0,0 +1,68 @@
+# The Tests / E2E Wasm workflow is used to run end-to-end tests on pull requests originating
+# from the ibc-go repository. The workflow specifically runs 08-wasm tests.
+name: Tests / E2E Wasm
+on:
+  workflow_dispatch:
+  pull_request:
+    paths:
+      - 'e2e/tests/wasm/grandpa_test.go' # the e2e test itself changes.
```
[the remark this answers is not in the document]
Comment (chatton) [reply .github/workflows/e2e-wasm.yaml:8]: yeah good call, no reason not to just include the wasm dir

File: .github/workflows/e2e-wasm.yaml [hunk 5]
```diff
@@ -0,0 +1,68 @@
+# The Tests / E2E Wasm workflow is used to run end-to-end tests on pull requests originating
+# from the ibc-go repository. The workflow specifically runs 08-wasm tests.
+name: Tests / E2E Wasm
+on:
+  workflow_dispatch:
+  pull_request:
+    paths:
+      - 'e2e/tests/wasm/**' # the e2e tests change.
+      - 'e2e/tests/wasm/contracts/**' # modifications to any contracts are made.
```
[the remark this answers is not in the document]
Comment (chatton) [reply .github/workflows/e2e-wasm.yaml:9]: you are correct :D 🦅 👁️

**Review by colin-axner — APPROVED** [review]
Thanks @chatton! :raised_hands:

**Comment by colin-axner** [comment]
looks like the pr needs to be rebased since the feature branch was merged

**Pull request merged**